File: CBORTest/BEncoding.cs
/*
Written in 2013 by Peter Occil.
Any copyright to this work is released to the Public Domain.
In case this is not possible, this work is also
licensed under the Unlicense: https://unlicense.org/

 */
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using PeterO.Cbor;
using PeterO.Numbers;

namespace PeterO {
  /// <summary>Contains methods for reading and writing objects
  /// represented in BEncode, a serialization format used in the
  /// BitTorrent protocol. For more information, see:
  /// http://wiki.theory.org/BitTorrentSpecification This class accepts
  /// BEncoded strings in UTF-8, and outputs BEncoded strings in UTF-8.
  /// This class also demonstrates how CBORObject supports predefined
  /// serialization formats.</summary>
  public static class BEncoding {
    private static void WriteUtf8(string s, Stream stream) {
      if (DataUtilities.WriteUtf8(s, stream, false) != 0) {
        throw new CBORException("invalid surrogate");
      }
    }

    private static CBORObject ReadDictionary(Stream stream) {
      var obj = CBORObject.NewMap();
      while (true) {
        int c = stream.ReadByte();
        if (c == 'e') {
          break;
        }
        CBORObject s = ReadString(stream, (char)c);
        CBORObject o = ReadObject(stream, false);
        obj[s] = o;
      }
      return obj;
    }

    private static CBORObject ReadInteger(Stream stream) {
      var builder = new StringBuilder();
      var start = true;
      while (true) {
        int c = stream.ReadByte();
        if (c < 0) {
          throw new CBORException("Premature end of data");
        }
        if (c is >= '0' and <= '9') {
          _ = builder.Append((char)c);
          start = false;
        } else if (c == 'e') {
          break;
        } else if (start && c == '-') {
          start = false;
          _ = builder.Append((char)c);
        } else {
          throw new CBORException("Invalid integer encoding");
        }
      }
      string s = builder.ToString();
      return s.Length >= 2 && s[0] == '0' && s[1] == '0' ?
        throw new CBORException("Invalid integer encoding") :
        s.Length >= 3 && s[0] == '-' && s[1] == '0' && s[2] == '0' ?
        throw new CBORException("Invalid integer encoding") :
        CBORObject.FromEInteger(
          EInteger.FromString(s));
    }

    private static CBORObject ReadList(Stream stream) {
      var obj = CBORObject.NewArray();
      while (true) {
        CBORObject o = ReadObject(stream, true);
        if (o == null) {
          break; // 'e' was read
        }
        _ = obj.Add(o);
      }
      return obj;
    }

    public static CBORObject Read(Stream stream) {
      return stream == null ? throw new
ArgumentNullException(nameof(stream)) : ReadObject(stream, false);
    }

    private static CBORObject ReadObject(Stream stream, bool allowEnd) {
      int c = stream.ReadByte();
      return c switch {
        'd' => ReadDictionary(stream),
        'l' => ReadList(stream),
        'i' => ReadInteger(stream),
        'e' => allowEnd ? null : throw new CBORException("Object expected"),
        '0' or '1' or '2' or '3' or '4' or '5' or '6' or '7' or '8' or '9'
=> ReadString(stream, (char)c),
        _ => throw new CBORException("Object expected"),
      };
    }

    private const string ValueDigits = "0123456789";

    public static string LongToString(long longValue) {
      return longValue == Int64.MinValue ?
        "-9223372036854775808" : longValue == 0L ?
        "0" : (longValue == Int32.MinValue) ? "-2147483648" :
EInteger.FromInt64(longValue).ToString();
    }

    private static CBORObject ReadString(Stream stream, char firstChar) {
      var builder = new StringBuilder();
      if (firstChar is < '0' or > '9') {
        throw new CBORException("Invalid integer encoding");
      }
      _ = builder.Append(firstChar);
      while (true) {
        int c = stream.ReadByte();
        if (c < 0) {
          throw new CBORException("Premature end of data");
        }
        if (c is >= '0' and <= '9') {
          _ = builder.Append((char)c);
        } else if (c == ':') {
          break;
        } else {
          throw new CBORException("Invalid integer encoding");
        }
      }
      string s = builder.ToString();
      if (s.Length >= 2 && s[0] == '0' && s[1] == '0') {
        throw new CBORException("Invalid integer encoding");
      }
      var numlength = EInteger.FromString(s);
      if (!numlength.CanFitInInt32()) {
        throw new CBORException("Length too long");
      }
      builder = new StringBuilder();
      return DataUtilities.ReadUtf8(
        stream,
        numlength.ToInt32Checked(),
        builder,
        false) switch {
        -2 => throw new CBORException("Premature end of data"),
        -1 => throw new CBORException("Invalid UTF-8"),
        _ => CBORObject.FromString(builder.ToString()),
      };
    }

    public static void Write(CBORObject obj, Stream stream) {
      if (obj == null) {
        throw new ArgumentNullException(nameof(obj));
      }
      if (obj.IsNumber) {
        if (stream == null) {
          throw new ArgumentNullException(nameof(stream));
        }
        stream.WriteByte(unchecked(0x69));
        WriteUtf8(obj.ToObject(typeof(EInteger)).ToString(), stream);
        stream.WriteByte(unchecked(0x65));
      } else if (obj.Type == CBORType.TextString) {
        string s = obj.AsString();
        long length = DataUtilities.GetUtf8Length(s, false);
        if (length < 0) {
          throw new CBORException("invalid string");
        }
        WriteUtf8(LongToString(length), stream);
        if (stream == null) {
          throw new ArgumentNullException(nameof(stream));
        }
        stream.WriteByte(unchecked((byte)':'));
        WriteUtf8(s, stream);
      } else if (obj.Type == CBORType.Map) {
        var hasNonStringKeys = false;
        foreach (CBORObject key in obj.Keys) {
          if (key.Type != CBORType.TextString) {
            hasNonStringKeys = true;
            break;
          }
        }
        if (hasNonStringKeys) {
          var valueSMap = new Dictionary<string, CBORObject>();
          // Copy to a map with String keys, since
          // some keys could be duplicates
          // when serialized to strings
          foreach (CBORObject key in obj.Keys) {
            CBORObject value = obj[key];
            string str = (key.Type == CBORType.TextString) ?
              key.AsString() : key.ToJSONString();
            valueSMap[str] = value;
          }
          if (stream == null) {
            throw new ArgumentNullException(nameof(stream));
          }
          stream.WriteByte(unchecked(0x64));
          foreach (KeyValuePair<string, CBORObject> entry in valueSMap) {
            string key = entry.Key;
            CBORObject value = entry.Value;
            long length = DataUtilities.GetUtf8Length(key, false);
            if (length < 0) {
              throw new CBORException("invalid string");
            }
            WriteUtf8(
              LongToString(length),
              stream);
            stream.WriteByte(unchecked((byte)':'));
            WriteUtf8(key, stream);
            Write(value, stream);
          }
          stream.WriteByte(unchecked(0x65));
        } else {
          if (stream == null) {
            throw new ArgumentNullException(nameof(stream));
          }
          stream.WriteByte(unchecked(0x64));
          foreach (CBORObject key in obj.Keys) {
            string str = key.AsString();
            long length = DataUtilities.GetUtf8Length(str, false);
            if (length < 0) {
              throw new CBORException("invalid string");
            }
            WriteUtf8(LongToString(length), stream);
            stream.WriteByte(unchecked((byte)':'));
            WriteUtf8(str, stream);
            Write(obj[key], stream);
          }
          stream.WriteByte(unchecked(0x65));
        }
      } else if (obj.Type == CBORType.Array) {
        if (stream == null) {
          throw new ArgumentNullException(nameof(stream));
        }
        stream.WriteByte(unchecked(0x6c));
        for (int i = 0; i < obj.Count; ++i) {
          Write(obj[i], stream);
        }
        stream.WriteByte(unchecked(0x65));
      } else {
        string str = obj.ToJSONString();
        long length = DataUtilities.GetUtf8Length(str, false);
        if (length < 0) {
          throw new CBORException("invalid string");
        }
        WriteUtf8(LongToString(length), stream);
        if (stream == null) {
          throw new ArgumentNullException(nameof(stream));
        }
        stream.WriteByte(unchecked((byte)':'));
        WriteUtf8(str, stream);
      }
    }
  }
}

File: src/com/upokecenter/test/BEncoding.java
package com.upokecenter.test; import com.upokecenter.util.*;
/*
Written in 2013 by Peter Occil.
Any copyright is dedicated to the Public Domain.
http://creativecommons.org/publicdomain/zero/1.0/
If you like this, you should donate to Peter O.
at: http://peteroupc.github.io/CBOR/
 */



import java.util.*;
import java.io.*;

    /**
     * Contains methods for reading and writing objects represented in
     * BEncode, a serialization format used in the BitTorrent protocol.
     * For more information, see: https://wiki.theory.org/BitTorrentSpecification#bencoding
     * This class accepts BEncoded strings in UTF-8, and outputs BEncoded
     * strings in UTF-8. This class also demonstrates how CBORObject supports
     * predefined serialization formats.
     * @param stream A readable data stream.
     */
  public final class BEncoding {
private BEncoding(){}
    private static void writeUtf8(String s, OutputStream stream) throws IOException {
      if (DataUtilities.WriteUtf8(s, stream, false) != 0)
        throw new CBORException("invalid surrogate");
    }
    private static CBORObject readDictionary(InputStream stream) throws IOException {
      CBORObject obj = CBORObject.NewMap();
      while (true) {
        int c = stream.read();
        if (c == 'e') {
          break;
        }
        CBORObject s = readString(stream, (char)c);
        CBORObject o = readObject(stream, false);
        obj.set(s,o);
      }
      return obj;
    }
    private static CBORObject readInteger(InputStream stream) throws IOException {
      StringBuilder builder = new StringBuilder();
      boolean start = true;
      while (true) {
        int c = stream.read();
        if (c < 0)
          throw new CBORException("Premature end of data");
        if (c >= (int)'0' && c <= (int)'9') {
          builder.append((char)c);
          start = false;
        } else if (c == (int)'e') {
          break;
        } else if (start && c == '-') {
          start = false;
          builder.append((char)c);
        } else {
          throw new CBORException("Invalid integer encoding");
        }
      }
      return CBORDataUtilities.ParseJSONNumber(
        builder.toString(), true, false, true);
    }
    private static CBORObject readList(InputStream stream) throws IOException {
      CBORObject obj = CBORObject.NewArray();
      while (true) {
        CBORObject o = readObject(stream, true);
        if (o == null) break;// 'e' was read
        obj.Add(o);
      }
      return obj;
    }
    public static CBORObject Read(InputStream stream) throws IOException {
      return readObject(stream, false);
    }
    private static CBORObject readObject(InputStream stream, boolean allowEnd) throws IOException {
      int c = stream.read();
      if (c == 'd')
        return readDictionary(stream);
      else if (c == 'l')
        return readList(stream);
      else if (allowEnd && c == 'e')
        return null;
      else if (c == 'i')
        return readInteger(stream);
      else if (c >= '0' && c <= '9') {
        return readString(stream, (char)c);
      } else {
        throw new CBORException("Object expected");
      }
    }
    private static CBORObject readString(InputStream stream, char firstChar) throws IOException {
      StringBuilder builder = new StringBuilder();
      if (firstChar < (int)'0' && firstChar > (int)'9') {
        throw new CBORException("Invalid integer encoding");
      } else {
        builder.append(firstChar);
      }
      while (true) {
        int c = stream.read();
        if (c < 0)
          throw new CBORException("Premature end of data");
        if (c >= (int)'0' && c <= (int)'9') {
          builder.append((char)c);
        } else if (c == (int)':') {
          break;
        } else {
          throw new CBORException("Invalid integer encoding");
        }
      }
      CBORObject number = CBORDataUtilities.ParseJSONNumber(
        builder.toString(), true, true, true);
      int length = 0;
      try {
        length = number.AsInt32();
      } catch (ArithmeticException ex) {
        throw new CBORException("Length too long", ex);
      }
      builder = new StringBuilder();
      switch (DataUtilities.ReadUtf8(stream, length, builder, false)) {
        case -2:
          throw new CBORException("Premature end of data");
        case -1:
          throw new CBORException("Invalid UTF-8");
        default:
          break;
      }
      return CBORObject.FromObject(builder.toString());
    }
    public static void Write(CBORObject obj, OutputStream stream) throws IOException {
      if (obj.getType() == CBORType.Number) {
        stream.write(((byte)((byte)'i')));
        writeUtf8(obj.AsBigInteger().toString(), stream);
        stream.write(((byte)((byte)'e')));
      } else if (obj.getType() == CBORType.TextString) {
        String s = obj.AsString();
        long length = DataUtilities.GetUtf8Length(s, false);
        if (length < 0)
          throw new CBORException("invalid String");
        writeUtf8(Long.toString((long)length), stream);
        stream.write(((byte)((byte)':')));
        writeUtf8(s, stream);
      } else if (obj.getType() == CBORType.Map) {
        boolean hasNonStringKeys = false;
        for(CBORObject key : obj.getKeys()) {
          if (key.getType() != CBORType.TextString) {
            hasNonStringKeys = true;
            break;
          }
        }
        if (hasNonStringKeys) {
          HashMap<String, CBORObject> sMap=new HashMap<String, CBORObject>();
          // Copy to a map with String keys, since
          // some keys could be duplicates
          // when serialized to strings
          for(CBORObject key : obj.getKeys()) {
            CBORObject value = obj.get(key);
            String str = (key.getType() == CBORType.TextString) ?
              key.AsString() : key.ToJSONString();
            sMap.put(str,value);
          }
          stream.write(((byte)((byte)'d')));
          for(Map.Entry<String, CBORObject> entry : sMap.entrySet()) {
            String key = entry.getKey();
            CBORObject value = entry.getValue();
            long length = DataUtilities.GetUtf8Length(key, false);
            if (length < 0)
              throw new CBORException("invalid String");
            writeUtf8(Long.toString((long)length), stream);
            stream.write(((byte)((byte)':')));
            writeUtf8(key, stream);
            Write(value, stream);
          }
          stream.write(((byte)((byte)'e')));
        } else {
          stream.write(((byte)((byte)'d')));
          for(CBORObject key : obj.getKeys()) {
            String str = key.AsString();
            long length = DataUtilities.GetUtf8Length(str, false);
            if (length < 0)
              throw new CBORException("invalid String");
            writeUtf8(Long.toString((long)length), stream);
            stream.write(((byte)((byte)':')));
            writeUtf8(str, stream);
            Write(obj.get(key), stream);
          }
          stream.write(((byte)((byte)'e')));
        }
      } else if (obj.getType() == CBORType.Array) {
        stream.write(((byte)((byte)'l')));
        for (int i = 0; i < obj.size(); i++) {
          Write(obj.get(i), stream);
        }
        stream.write(((byte)((byte)'e')));
      } else {
        String str = obj.ToJSONString();
        long length = DataUtilities.GetUtf8Length(str, false);
        if (length < 0)
          throw new CBORException("invalid String");
        writeUtf8(Long.toString((long)length), stream);
        stream.write(((byte)((byte)':')));
        writeUtf8(str, stream);
      }
    }
  }
